### PR TITLE
refactor: switch to tls cipher suite in stdlib

### LINKED ISSUE #88547: Switch to go1.14 tls cipher suite helper methods
Once we upgrade to go1.14, we can use the new CipherSuites and InsecureCipherSuites functions and CipherSuiteName function to build the supported cipher suite flags.

- https://golang.org/pkg/crypto/tls/#CipherSuites
- https://golang.org/pkg/crypto/tls/#InsecureCipherSuites
- https://golang.org/pkg/crypto/tls/#CipherSuiteName

/sig api-machinery

### LINKED ISSUE #105063: k8s control plane components support insecure cipher suites
#### What happened:
Nessus tool scan reports insecure cipher suites is supported by the kube-apiserver, kube-controller-manager and kube-scheduler components.

#### What you expected to happen:
Should not have any vulnerabilities reported in nessus tool scan.

#### How to reproduce it (as minimally and precisely as possible):
- kubernetes cluster is installed and is in operational state
- Setup and scan for vulnerabilities as suggested at https://docs.tenable.com/nessus/8_14/Content/GetStarted.htm

#### Anything else we need to know?:
Nessus tool version: 8.14.0(#261) LINUX

Nessus report:
```
1. HIGH
SSL Medium Strength Cipher Suites Supported (SWEET32)

Description
The remote host supports the use of SSL ciphers that offer medium strength encryption. Nessus regards medium strength as any encryption that uses key lengths at least 64 bits and less than 112 bits, or else that uses the 3DES encryption suite.

Note that it is considerably easier to circumvent medium strength encryption if the attacker is on the same physical network.
Solution
Reconfigure the affected application if possible to avoid use of medium strength ciphers.
See Also
https://www.openssl.org/blog/blog/2016/08/24/sweet32/
https://sweet32.info
Output
•	  Medium Strength Ciphers (> 64-bit and < 112-bit key, or 3DES)
•	
•	    Name                          Code             KEX           Auth     Encryption             MAC
•	    ----------------------        ----------       ---           ----     ---------------------  ---
•	    ECDHE-RSA-DES-CBC3-SHA        0xC0, 0x12       ECDH          RSA      3DES-CBC(168)          SHA1
•	    DES-CBC3-SHA                  0x00, 0x0A       RSA           RSA      3DES-CBC(168)          SHA1
•	
•	The fields above are :
•	
•	  {Tenable ciphername}
•	  {Cipher ID code}
•	  Kex={key exchange}
•	  Auth={authentication}
•	  Encrypt={symmetric encryption method}
•	  MAC={message authentication code}
•	  {export flag}
•	
Port	Hosts
10257 / tcp / www	10.24.x.x
6443 / tcp / www	10.24.x.x
10259 / tcp / www	10.24.x.x
```

#### Environment:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"clean", BuildDate:"2020-06-26T03:47:41Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"clean", BuildDate:"2020-06-26T03:39:24Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="18.04.4 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.4 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`):
`Linux k8s-master 4.15.0-135-generic #139-Ubuntu SMP Mon Jan 18 17:38:24 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by knight42**
Signed-off-by: Jian Zeng <zengjian.zj@bytedance.com>

#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #88547 
xref #105063 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go [hunk 1]
```diff
@@ -23,39 +23,24 @@ import (
...
+	for _, suite := range tls.CipherSuites() {
+		ciphers[suite.Name] = suite.ID
+	}
+	// keep legacy names for backward compatibility
+	ciphers["TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305"] = tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+	ciphers["TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305"] = tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
+
+	for _, suite := range tls.InsecureCipherSuites() {
+		insecureCiphers[suite.Name] = suite.ID
+	}
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:43]: now that we're using these, we can drop the `TestConstantMaps` unit test that ensured we stayed in sync with the stdlib

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go [hunk 2]
```diff
@@ -23,39 +23,24 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-// ciphers maps strings into tls package cipher constants in
-// https://golang.org/pkg/crypto/tls/#pkg-constants
-// to be replaced by tls.CipherSuites() when the project migrates to go1.14.
-var ciphers = map[string]uint16{
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:29]: add a test case to TestStrToUInt16 that specifies all of the existing secure and insecure strings, ensure it passes on master and on this PR to make sure we're not dropping/changing any mappings

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go [hunk 3]
```diff
@@ -23,39 +23,24 @@ import (
...
-	"TLS_ECDHE_ECDSA_WITH_RC4_128_SHA":        tls.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,
-	"TLS_ECDHE_RSA_WITH_RC4_128_SHA":          tls.TLS_ECDHE_RSA_WITH_RC4_128_SHA,
-	"TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256": tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,
-	"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256":   tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,
+func init() {
+	for _, suite := range tls.CipherSuites() {
+		ciphers[suite.Name] = suite.ID
+	}
+	// keep legacy names for backward compatibility
+	ciphers["TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305"] = tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:38]: this doesn't match the previous cipher, right?

```
"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305": tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
```

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go [hunk 4]
```diff
@@ -23,39 +23,24 @@ import (
...
+	ciphers["TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305"] = tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:39]: this doesn't match the previous mapping, right?
```
"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305": tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
```

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go (hunk 3 above)
Comment (knight42) [reply staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:38]: https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/crypto/tls/cipher_suites.go;l=687

`tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305` is actually the same as `tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256`, `tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305` is retained for backwards compability.

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go (hunk 4 above)
Comment (knight42) [reply staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:39]: `tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305` is also the same as `tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256`

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag_test.go [hunk 5]
```diff
@@ -61,55 +58,74 @@ func TestStrToUInt16(t *testing.T) {
...
+				"TLS_CHACHA20_POLY1305_SHA256",
+				"TLS_AES_256_GCM_SHA384",
+
+				"TLS_RSA_WITH_RC4_128_SHA",
+				"TLS_RSA_WITH_AES_128_CBC_SHA256",
+				"TLS_ECDHE_ECDSA_WITH_RC4_128_SHA",
+				"TLS_ECDHE_RSA_WITH_RC4_128_SHA",
+				"TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
+				"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256",
+			},
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag_test.go:90]: add
```
"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256",
```

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag_test.go [hunk 6]
```diff
@@ -61,55 +58,74 @@ func TestStrToUInt16(t *testing.T) {
...
+				tls.TLS_CHACHA20_POLY1305_SHA256,
+				tls.TLS_AES_256_GCM_SHA384,
+
+				tls.TLS_RSA_WITH_RC4_128_SHA,
+				tls.TLS_RSA_WITH_AES_128_CBC_SHA256,
+				tls.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,
+				tls.TLS_ECDHE_RSA_WITH_RC4_128_SHA,
+				tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,
+				tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,
+			},
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag_test.go:118]: add
```
tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
```

File: staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:39]: add the _114 mappings to the unit test to verify this

Comment (knight42) [reply staging/src/k8s.io/component-base/cli/flag/ciphersuites_flag.go:39]: fixed

**Comment by knight42** [comment]
@liggitt Hi! The commits have been squashed now, could I get a `/approve`?

**Pull request merged**